Mathlib/CategoryTheory/Limits/Shapes/StrictInitial.lean

```diff
@@ -0,0 +1,283 @@
+Strictly speaking, this says that *any* initial object must be strict, rather than that strict
+initial objects exist, which turns out to be a more useful notion to formalise.
+
+If the binary product of `X` with a strict initial object exists, it is also initial.
+
+To show a category `C` with an initial object has strict initial objects, the most convenient way
+is to show any morphism to the (chosen) initial object is an isomorphism and use
+`has_strict_initial_objects_of_initial_is_strict`.
+
+The dual notion (strict terminal objects) occurs much less frequently in practice so is ignored.
+
+## TODO
+
+* Construct examples of this: `Type*`, `Top`, `Groupoid`, simplicial types, posets.
+* Construct the bottom element of the subobject lattice given strict initials.
+* Show cartesian closed categories have strict initials
+
+## References
+* https://ncatlab.org/nlab/show/strict+initial+object
+-/
+
+
+universe v u
+
+namespace CategoryTheory
+
+namespace Limits
+
+open Category
+
+variable (C : Type u) [Category.{v} C]
+
+section StrictInitial
+
+/-- We say `C` has strict initial objects if every initial object is strict, ie given any morphism
+`f : A ⟶ I` where `I` is initial, then `f` is an isomorphism.
+
+Strictly speaking, this says that *any* initial object must be strict, rather than that strict
+initial objects exist.
+-/
+class HasStrictInitialObjects : Prop where
+  out : ∀ {I A : C} (f : A ⟶ I), IsInitial I → IsIso f
+#align category_theory.limits.has_strict_initial_objects CategoryTheory.Limits.HasStrictInitialObjects
+
+variable {C}
+
+section
+
+variable [HasStrictInitialObjects C] {I : C}
+
+theorem IsInitial.isIso_to (hI : IsInitial I) {A : C} (f : A ⟶ I) : IsIso f :=
+  HasStrictInitialObjects.out f hI
+#align category_theory.limits.is_initial.is_iso_to CategoryTheory.Limits.IsInitial.isIso_to
+
+theorem IsInitial.strict_hom_ext (hI : IsInitial I) {A : C} (f g : A ⟶ I) : f = g := by
+  haveI := hI.isIso_to f
+  haveI := hI.isIso_to g
+  exact eq_of_inv_eq_inv (hI.hom_ext (inv f) (inv g))
+#align category_theory.limits.is_initial.strict_hom_ext CategoryTheory.Limits.IsInitial.strict_hom_ext
+
+theorem IsInitial.subsingleton_to (hI : IsInitial I) {A : C} : Subsingleton (A ⟶ I) :=
+  ⟨hI.strict_hom_ext⟩
+#align category_theory.limits.is_initial.subsingleton_to CategoryTheory.Limits.IsInitial.subsingleton_to
+
+instance (priority := 100) initial_mono_of_strict_initial_objects : InitialMonoClass C
+    where isInitial_mono_from := @fun _I _A hI =>
+    { right_cancellation := @fun _B _g _h _i => hI.strict_hom_ext _ _ }
+#align category_theory.limits.initial_mono_of_strict_initial_objects CategoryTheory.Limits.initial_mono_of_strict_initial_objects
+
+/-- If `I` is initial, then `X ⨯ I` is isomorphic to it. -/
+@[simps! hom]
+noncomputable def mulIsInitial (X : C) [HasBinaryProduct X I] (hI : IsInitial I) : X ⨯ I ≅ I := by
+   have h: X ⨯ I ⟶ I := prod.snd
+   have := hI.isIso_to h
+   exact asIso h
```

Comment:
added those have in, because the type inference failed otherwise, also below. Please check if it is ok like this or if there is a better way

Reply:
Does
```suggestion
   have := hI.isIso_to (prod.snd : X ⨯ I ⟶ I)
   exact asIso (prod.snd : X ⨯ I ⟶ I)
```
work?

Note that `have` is for props, so please use `let` when working with data.